File: leetcode/hard/java/lc42.java
class Solution {
    public int trap(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int leftMax, rightMax, maxArea;
        leftMax = rightMax = maxArea = 0;
        while (left < right) {
            if (height[left] < height[right]) {
                if (height[left] >= leftMax) {
                    leftMax = height[left];
                } else {
                    maxArea += leftMax - height[left];
                }
                left++;
            } else {
                if (height[right] >= rightMax) {
                    rightMax = height[right];
                } else {
                    maxArea += rightMax - height[right];
                }
                right--;
            }
        }
        return maxArea;
    }
}

File: leetcode/hard/python/lc_42.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        left_max, right_max, max_area = 0, 0, 0
        while left < right:
            if height[left] < height[right]:
                if height[left] >= left_max:
                    left_max = height[left]
                else:
                    max_area += left_max - height[left]
                left += 1
            else:
                if height[right] >= right_max:
                    right_max = height[right]
                else:
                    max_area += right_max - height[right]
                right -= 1
        return max_area
